// dm-form/select/index.js
// Generated by CoffeeScript 1.8.0
(function() {
  var FormSelect,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  module.exports = FormSelect = (function(_super) {
    __extends(FormSelect, _super);

    function FormSelect() {
      return FormSelect.__super__.constructor.apply(this, arguments);
    }

    FormSelect.prototype.view = __dirname;

    FormSelect.prototype.style = __dirname;

    FormSelect.prototype.name = 'dm-form:select';

    FormSelect.prototype._getExtraFields = function() {
      return ['options'];
    };

    FormSelect.prototype.init = function() {
      return FormSelect.__super__.init.apply(this, arguments);
    };

    FormSelect.prototype.optionValue = function(option) {
      if ((option != null ? option.value : void 0) != null) {
        return option.value;
      } else {
        return this.optionName(option);
      }
    };

    FormSelect.prototype.optionName = function(option) {
      if ((option != null ? option.name : void 0) != null) {
        return option.name;
      } else if ((option != null ? option.title : void 0) != null) {
        return option.title;
      } else if (option != null ? option.content : void 0) {
        return option.content;
      } else if ((option != null ? option.value : void 0) != null) {
        return option.value;
      } else {
        return option;
      }
    };

    return FormSelect;

  })(require('../input'));

}).call(this);
