// dm-form/text/index.js
// Generated by CoffeeScript 1.8.0
(function() {
  var FormText,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  module.exports = FormText = (function(_super) {
    __extends(FormText, _super);

    function FormText() {
      return FormText.__super__.constructor.apply(this, arguments);
    }

    FormText.prototype.view = __dirname;

    FormText.prototype.style = __dirname;

    FormText.prototype.name = 'dm-form:text';

    FormText.prototype._getExtraFields = function() {
      return ['lazy'];
    };

    FormText.prototype.init = function() {
      FormText.__super__.init.apply(this, arguments);
      if (this.model.get('name') === '$id') {
        this.setAttribute('lazy', true);
      }
      if (!this.getAttribute('lazy')) {
        return this.model.ref('_value', 'value');
      }
    };

    FormText.prototype.lazySet = function() {
      if (!this.getAttribute('lazy')) {
        return;
      }
      return this.model.setDiff('value', this.model.get('_value'));
    };

    return FormText;

  })(require('../input'));

}).call(this);
