// dm-form/textarea/index.js
// Generated by CoffeeScript 1.8.0
(function() {
  var FormTextarea,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  module.exports = FormTextarea = (function(_super) {
    __extends(FormTextarea, _super);

    function FormTextarea() {
      return FormTextarea.__super__.constructor.apply(this, arguments);
    }

    FormTextarea.prototype.view = __dirname;

    FormTextarea.prototype.style = __dirname;

    FormTextarea.prototype.name = 'dm-form:textarea';

    FormTextarea.prototype._getExtraFields = function() {
      return ['lazy'];
    };

    FormTextarea.prototype.init = function() {
      FormTextarea.__super__.init.apply(this, arguments);
      if (this.model.get('name') === '$id') {
        this.setAttribute('lazy', true);
      }
      if (!this.getAttribute('lazy')) {
        return this.model.ref('_value', 'value');
      }
    };

    FormTextarea.prototype.lazySet = function() {
      if (!this.getAttribute('lazy')) {
        return;
      }
      return this.model.setDiff('value', this.model.get('_value'));
    };

    return FormTextarea;

  })(require('../input'));

}).call(this);
