// derby-webpack/gulp/base.js
// Generated by CoffeeScript 1.9.3
(function() {
  var DeepMerge, deepmerge;

  DeepMerge = require('deep-merge');

  deepmerge = DeepMerge(function(target, source, key) {
    if (target instanceof Array) {
      return [].concat(target, source);
    }
    return source;
  });

  module.exports = function(options) {
    var config;
    config = {
      module: {
        loaders: [
          {
            test: /\.jade$/,
            loaders: [__dirname + ("/../loaders/derby-jade-loader?" + (options.moduleMode ? 'module' : ''))]
          }, {
            include: /\.coffee$/,
            loader: "coffee",
            exclude: options.componentsPaths
          }, {
            include: options.componentsPaths,
            loaders: [__dirname + '/../loaders/derby-component-loader', 'coffee']
          }, {
            include: /\.json$/,
            loader: 'json'
          }, {
            include: /\.yaml$/,
            loader: 'json!yaml'
          }
        ]
      },
      resolveLoader: {
        root: __dirname + '/../node_modules'
      },
      resolve: {
        extensions: ['', '.json', '.js', '.yaml', '.coffee']
      }
    };
    if (process.env.NODE_ENV !== 'production') {
      config.devtool = 'source-map';
      config.debug = true;
    }
    return {
      config: function(overrides) {
        return deepmerge(config, overrides || {});
      },
      onBuild: function(done) {
        return function(err, stats) {
          if (err) {
            console.log('Error', err);
          } else {
            console.log(stats.toString());
          }
          return typeof done === "function" ? done() : void 0;
        };
      }
    };
  };

}).call(this);
