// derby-webpack/gulp/index.js
// Generated by CoffeeScript 1.9.3
(function() {
  var _, gulp, nodemon, path;

  gulp = require('gulp');

  nodemon = require('nodemon');

  path = require('path');

  _ = require('lodash');

  module.exports = function(options) {
    if (options == null) {
      options = {};
    }
    _.defaults(options, {
      dirname: path.dirname(path.dirname(path.dirname(__dirname))),
      frontend: {},
      backend: {},
      webpackPort: 3010,
      serverPort: 3000,
      moduleMode: false,
      componentsPaths: [/\/components\/.+\.coffee$/, /\/pages\/.+\.coffee$/],
      apps: ['app']
    });
    require('./frontend')(options);
    require('./backend')(options);
    gulp.task('build', ['frontend-build', 'backend-build']);
    gulp.task('watch', ['frontend-watch', 'backend-watch']);
    return gulp.task('run', ['frontend-watch', 'backend-watch'], function() {
      return nodemon({
        execMap: {
          js: 'node'
        },
        script: options.dirname + '/build/backend',
        ignore: ['*'],
        watch: ['foo/'],
        ext: 'noop'
      }).on('restart', function() {});
    });
  };

}).call(this);
